File: 20-valid-parentheses/valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        stack = []
        mapping = {')': '(', ']': '[', '}': '{'}

        for char in s:
            if char in mapping.values():
                stack.append(char)  # opening bracket
            elif char in mapping:
                if not stack or stack[-1] != mapping[char]:
                    return False  # mismatch
                stack.pop()  # matched pair
            else:
                return False  # invalid character (optional)

        return not stack  # stack must be empty if valid

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack=new Stack();
        Map<Character,Character> mapping=new HashMap();
        mapping.put(')','(');
        mapping.put(']','[');
        mapping.put('}','{');
      
        for(char ch:s.toCharArray()){
            if (mapping.containsValue(ch)){
                stack.push(ch);
            }else if (mapping.containsKey(ch)){

            if(stack.isEmpty() || stack.peek()!=mapping.get(ch)){
                return false;
            }
            stack.pop();
            }else{
                return false; //invalid character
            }
        }
        return stack.isEmpty();
    }
}